.circleci/config.yml

```
@@ -208,9 +208,44 @@ commands:
 
   run_build:
     steps:
+      # TODO: remove the line below
+      # Before merging this PR we need to use an alternative develop branch to test the functionality, `develop-cache-test`
+
+      # On `develop-cache-test` we intentionally do NOT restore any ccache, so every run starts from scratch.
+      # Other branches restore the most recent cache produced on `develop-cache-test`.
+      - when:
+          condition:
+            not:
+              equal: ["develop-cache-test", << pipeline.git.branch >>]
+          steps:
+            - restore_cache:
+                keys:
+                  # Reuse cache produced on develop-cache-test (prefix match; restores the most recent).
+                  - v1-ccache-{{ arch }}-develop-cache-test-
+                  - v1-ccache-{{ arch }}-
       - run:
           name: Build
-          command: scripts/ci/build.sh
+          command: |
+            export CCACHE_DIR="$HOME/.ccache"
+            export CCACHE_BASEDIR="$(pwd)"
+            export CCACHE_NOHASHDIR=1
+            mkdir -p "$CCACHE_DIR"
+            export CMAKE_OPTIONS="${CMAKE_OPTIONS:-} -DCMAKE_C_COMPILER_LAUNCHER=ccache -DCMAKE_CXX_COMPILER_LAUNCHER=ccache"
+            ccache -z || true
+
+            scripts/ci/build.sh
+
+            ccache -s || true
+
+      # On `develop-cache-test` we always generate a fresh cache by never restoring and by saving under a unique key.
+      - when:
+          condition:
+            equal: ["develop-cache-test", << pipeline.git.branch >>]
+          steps:
+            - save_cache:
+                key: v1-ccache-{{ arch }}-develop-cache-test-{{ .Revision }}
```

I expect that we might want to invalidate cache whenever we suspect something weird is going on. It would be convenient to be able to do that without having to merge a PR. In an extreme case we might even not be able to merge one due to required jobs failing because of cache. In addition to the hard-coded version we could also keep a variable in CircleCI config that we can bump to get the same effect:

```suggestion
                key: v1-{{ .Environment.CCACHE_KEY }}-ccache-{{ arch }}-develop-cache-test-{{ .Revision }}
```